src/core/or/policies.c

```diff
@@ -495,8 +500,39 @@ fascist_firewall_prefer_ipv6_impl(const or_options_t *options)
 MOCK_IMPL(int,
 fascist_firewall_rand_prefer_ipv6_addr, (void))
 {
-  /* TODO: Check for failures, and infer our preference based on this. */
-  return crypto_rand_int(2);
+  or_state_t *state = get_or_state();
+  time_t current_time = time(NULL);
+
+  /* Clear IPv4 and IPv6 failure entries if they are more than a week old. */
+  if (state->IPv4AutoFailTime &&
```

Comment:
Maybe we should track this information over smaller time periods, so we can forget the earliest failures more regularly. rephist.c has some code for tracking state over multiple time periods.